### PR TITLE
TypeError while hashing and some Windows compatibility

### LINKED ISSUE #708: ESO retrieve_data TypeError
I wanted to download some data from the ESO archive and I've followed the example before I try to get my data but unfortunatly I hit some TypeError:

```
>>> from astroquery.eso import Eso
>>> eso = Eso()
>>> eso.login("xxx")
INFO: Authenticating xxx on www.eso.org... [astroquery.eso.core]
INFO: Authentication successful! [astroquery.eso.core]
>>> table = eso.query_instrument('midi', column_filters={'target':'NGC 4151', 'stime':'2007-01-01', 'etime':'2008-01-01'}, columns=['night'])
>>> data_files = eso.retrieve_data(table['DP.ID'][:2])
INFO: Staging request... [astroquery.eso.core]

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-3-54bc744a4130> in <module>()
----> 1 data_files = eso.retrieve_data(table['DP.ID'][:2])

C:\Programming\Anaconda\envs\-\lib\site-packages\astroquery-0.3.2-py3.5.egg\astroquery\eso\core.py in retrieve_data(self, datasets, cache)
    583                 # should be included too
    584                 data_download_form = self._activate_form(
--> 585                     data_confirmation_form, form_index=-1)
    586                 log.info("Staging form is at {0}."
    587                          .format(data_download_form.url))

C:\Programming\Anaconda\envs\-\lib\site-packages\astroquery-0.3.2-py3.5.egg\astroquery\eso\core.py in _activate_form(self, response, form_index, inputs, cache, method)
    167             response = self._request("POST", url, files=payload, cache=cache)
    168         elif fmt == 'application/x-www-form-urlencoded':
--> 169             response = self._request("POST", url, data=payload, cache=cache)
    170 
    171         return response

C:\Programming\Anaconda\envs\-\lib\site-packages\astroquery-0.3.2-py3.5.egg\astroquery\query.py in _request(self, method, url, params, data, headers, files, save, savedir, timeout, cache, stream, auth)
    171                                              auth=auth)
    172             else:
--> 173                 response = query.from_cache(self.cache_location)
    174                 if not response:
    175                     response = query.request(self._session,

C:\Programming\Anaconda\envs\-\lib\site-packages\astroquery-0.3.2-py3.5.egg\astroquery\query.py in from_cache(self, cache_location)
     78 
     79     def from_cache(self, cache_location):
---> 80         request_file = self.request_file(cache_location)
     81         try:
     82             with open(request_file, "rb") as f:

C:\Programming\Anaconda\envs\-\lib\site-packages\astroquery-0.3.2-py3.5.egg\astroquery\query.py in request_file(self, cache_location)
     74 
     75     def request_file(self, cache_location):
---> 76         fn = os.path.join(cache_location, self.hash() + ".pickle")
     77         return fn
     78 

C:\Programming\Anaconda\envs\-\lib\site-packages\astroquery-0.3.2-py3.5.egg\astroquery\query.py in hash(self)
     63                     request_key += (tuple(sorted(k.items())),)
     64                 elif isinstance(k, tuple) or isinstance(k, list):
---> 65                     request_key += (tuple(sorted(k)),)
     66                 elif k is None:
     67                     request_key += (None,)

TypeError: unorderable types: NoneType() < str()
```

At first I thought it was a problem regarding `gzip` but after editing the source to use `7z` the error persisteted. The problem is that the `self.data` in the `query.hash` contains:

```
[('dataset', 'SAF+MIDI.2007-02-07T07:01:51.000'), ('dataset', 'SAF+MIDI.2007-02-07T07:02:49.000'), 
('requestDescription', None), 
('deliveryMediaType', 'WEB'), ('deliveryMediaType', None), 
('requestCommand', 'on'), ('requestCommand', None), ('requestCommand', None), ('requestCommand', 'SELECTIVE_HOTFLY'), 
('submit', 'Submit')]
```

containing 2 `deliveryMediaType` (one `None`) and 4 `requestCommand` (2 of them `None`) which trigger the error.

I'm using Windows 10, Python 3.5 and astroquery 0.3.2 (lxml 3.6 and requests 2.10)

The astroquery 0.3.1 doesn't fail **here** (it still fails but because the downloaded files contain "forbidden characters" for windows files `:`).

### PR DESCRIPTION
**Pull request by MSeifert04**
Ignores "None" while hashing. (Fixes #708)

And something for windows compatibility:
Fallback to "7z" for unpacking if "gzip" is not avaiable.
Replace forbidden windows characters in local_filename for query.

I feel that my implementation isn't very good - these were just the changes I needed to get the ESO file request running.

### CODE REVIEW

**Comment by MSeifert04** [comment]
@keflavich I've scanned through `os.path` and `pathlib` and found no such function. This issue is a bit convolved since it's a filesystem and/or os issue and there are [several reserved characters](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx) for ntfs or windows. I wouldn't be surprised if there is some general way but I haven't found it.

I just fixed the part that stopped me from working with the eso-archive: The `:` in the filenames resulting in an `OSError: [Errno 22] ...`

Actually I'm not very satisfied with any part of this PR (the sorting is overly complicated/inefficient, the 7z or gzip choice is a `try ... except`-loop). So if anyone wants to create some better alternative please go ahead. The PR was mostly meant as a starting point that worked for me.

**Comment by keflavich** [comment]
OK, thanks.  The PR is broadly fine, and I guess the solution you came up with is good for this specific case since ESO files should have a specific filename.

Can we enable appveyor tests to test this on windows machines?
@bsipocz @astrofrog

**Comment by MSeifert04** [comment]
Is it possible that besides Windows also Mac doesn't allow the `:` in filenames?

**Comment by keflavich** [comment]
No, mac allows them.  They're annoying because they have to be escaped on the command line, but mac is just unix and can live with any ascii character (afaict)

**Comment by MSeifert04** [comment]
ok, I was just reading through https://kb.acronis.com/content/39790 and there it says it's the only forbidden character on OS 9 and OS X.

**Comment by keflavich** [comment]
Interesting article.  It's wrong on two counts.  I had no trouble making a file with `:` in it, and files beginning with `.` are built in to the OS (configuration files in the home directory, for example).

Anyway, let me know when you think this PR is ready.   I'm willing to merge it, though it would be better if we could set up useful tests.

**Comment by MSeifert04** [comment]
That's really weird is was highly ranked on google searching for "file system illegal characters". But doesn't matter.

I don't have time this week and next so maybe this is ready to go. In theory I was hoping to find the time to write a function containing a regex to replace all illegal windows characters from a filename and optimize that key-function a little but both are not strictly necessary (for now). I noticed I forgot the `break` in the `gzip or 7-zip`-loop.

Testing doesn't really work locally, I had 10 failures in the first ~30 tests (and of those ~10 were skipped). I think `astroquery` just doesn't fully support Windows at this point. I could create a gist containing the failures but AppVeyor probably won't run. 

Also the ESO-request needs a login and I don't plan to create a test containing my login-data, so I don't know how to create a regression test for the `None`-failure there #708.

TL;DR;
I don't know how to create tests for the fixed failures and don't have the time currently to make the code "cleaner".

**Comment by keflavich** [comment]
@MSeifert04 Thanks.  Could you post the windows failures as a gist here?  I think I'll just merge this afterward, unless @bsipocz has any further concerns or suggestions.

**Comment by bsipocz** [comment]
I've opened a PR to enable appveyor. We will have plenty of failures I think, but as this PR fixes some of those I would like to merge this now so have a better picture of the job ahead.

**Comment by bsipocz** [comment]
Thanks @MSeifert04!
